UHM-7475, configure person relationships widget with a specific type

cioan opened the pull request.

mseaton commented on omod/src/main/webapp/resources/scripts/field/registerPersonRelationship.js:
```diff
@@ -0,0 +1,43 @@
+            limit: 10,
+            q: request.term,
+            v: 'custom:(id,uuid,display,gender,age)'
+          }, function (result) {
```
It looks to me like you are configuring this widget with a `gender` property, but I don't see where you are actually limiting the results to persons that match this gender.  Shouldn't that be done in a filter here?

mseaton commented on api/src/main/java/org/openmrs/module/registrationapp/model/RegisterPersonRelationshipWidget.java:
```diff
@@ -0,0 +1,79 @@
+
+import org.codehaus.jackson.annotate.JsonProperty;
+
+public class RegisterPersonRelationshipWidget extends Widget{
```
Nitpick, but if you can run a formatter on this file to clean up the spacing before/after curly braces, between methods, between properties, etc that would be good.  It seems a bit inconsistent.

mseaton reviewed (commented):
Overall looks fine to me, would be good to see a screenshot / animated gif that demonstrates what this looks like.  A few small comments to address.

mseaton commented on api/src/main/java/org/openmrs/module/registrationapp/model/RegisterPersonRelationshipWidget.java:
```diff
@@ -0,0 +1,79 @@
+        this.config = config;
+    }
+
+    public static class Config {
```
See my comments on the pihcore PR.  Basically, let's make this a top-level class named `PersonRelationshipConfig` and also add the `label` property you have on the config in the other PR.

mogoodrich replied:
@mseaton  I would think it makes sense to keep this like it is, because it follows the existing pattern used for the other widgets?  Or am I missing something?

mseaton replied:
Ok, fine if that's what we've done...

mogoodrich commented on api/src/main/java/org/openmrs/module/registrationapp/model/RegisterPersonRelationshipWidget.java:
```diff
@@ -0,0 +1,79 @@
+
+    public static class Config {
+        @JsonProperty
+        private String type;
```
What does type property refer to?

mogoodrich commented on api/src/main/java/org/openmrs/module/registrationapp/model/RegisterPersonRelationshipWidget.java:
```diff
@@ -0,0 +1,79 @@
+        private String relationshipType;
+
+        @JsonProperty
+        private String gender;
```
this limits the people returns by gender, if specified?  (makes sense, as long as null is allowed, just clarifying)

mogoodrich reviewed (commented):
Thanks @cioan !  Looks like the right approach to me.  Agree with Mike about gender, but assume that this is because this is a work-in-progress and hasn't been implemented.

I'm guessing the "type" property is to specific whether it is an "a_is_to_b" or "b_is_to_a" relationship, which we will need when specifying whether you are selecting the "Parent' or "Child" in a "Parent/Child" relationship.

But, yes, looks like the right approach in general!

cioan commented:
Thank you @mseaton and @mogoodrich !  I have implemented the gender filtering:
<img width="931" alt="Screenshot 2023-10-18 at 10 17 23 AM" src="https://github.com/openmrs/openmrs-module-registrationapp/assets/1633285/19110074-e61e-4186-8231-05f11553fdf7">

I also removed the "type" property and reformatted the code. I kept the inner Config class per @mogoodrich comment that this is the current pattern implemented for all the other widgets. Please let me know if you would like to discuss this further.

cioan commented:
I have updated the widget so the direction of the relationship could be configured. And, therefore, now we are able to configure the widget to record a child of the patient:
<img width="931" alt="Screenshot 2023-10-19 at 9 33 18 AM" src="https://github.com/openmrs/openmrs-module-registrationapp/assets/1633285/56ecff27-7595-4683-abfa-0d2968c7b25a">
<img width="931" alt="Screenshot 2023-10-19 at 9 34 28 AM" src="https://github.com/openmrs/openmrs-module-registrationapp/assets/1633285/e9980b58-fcdc-4eda-95c9-2e70df4741c4">

<img width="931" alt="Screenshot 2023-10-19 at 9 35 03 AM" src="https://github.com/openmrs/openmrs-module-registrationapp/assets/1633285/a8f4fb84-45dd-42fd-b1e9-befa1be78fc7">

cioan commented:
Hi @mseaton and @mogoodrich , in the latest commit, I added the capability to record multiple relationships of the same type by leveraging the "multipleValues" configuration property:
![Screenshot 2023-10-20 at 9 14 32 AM](https://github.com/openmrs/openmrs-module-registrationapp/assets/1633285/41e792db-c622-48c6-a3a5-00e3333f9ec2)
![Screenshot 2023-10-20 at 9 15 02 AM](https://github.com/openmrs/openmrs-module-registrationapp/assets/1633285/170dc66a-2591-45a3-b0a0-28e0975a7767)

![Screenshot 2023-10-20 at 9 16 34 AM](https://github.com/openmrs/openmrs-module-registrationapp/assets/1633285/437b523c-70d9-4ec8-88fb-53e8774b841a)

Please let me know what you think. I still need to implement the removeRelationship function and to figure out the display Template for the Confirm page. But now, we are able to record multiple children of a patient using the registration form.

mseaton commented:
Screenshots look great to me @cioan !

cioan commented:
Thanks @mseaton ! I am going to merge in all those pending PRs so they have time to propagate to test servers over the weekend.
